File: rust/api/src/descriptions.rs
//! `/descriptions` — the shared names transactions are filed under.
//!
//! Ported from `routes/descriptions.ts`.
//!
//! # A description is a name many transactions share
//!
//! So renaming one is not a single-row edit: every transaction pointing at it
//! carries a copy of the name, and all of them move together. Merging is the
//! same operation with a second step — repoint the rows, then remove the name
//! nothing references any more.
//!
//! Both go through `ledger_update` rather than a bulk `UPDATE`. No hook reads
//! `descriptionId`, so this is not about side effects; it is that `name` moves
//! with it, and a gate that some name-changes skip is a gate with an exception
//! nobody will remember.
//!
//! # Uniqueness is case-insensitive, and deliberately
//!
//! "Amazon" and "amazon" are the same merchant. Allowing both is how a
//! description list grows a dozen spellings of one shop, which is the thing
//! this feature exists to prevent.

use crate::id::{cuid, now_iso};
use crate::{ApiError, Path, Response};
use avoir_db::ledger::{ledger_update, LedgerUpdate};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use sqlx::SqlitePool;

/// A description, as all four of its routes return it.
///
/// Four separate `json!({ "id": …, "name": … })` literals built this shape
/// before — create, update, read and list — with nothing connecting them. See
/// `budgets.rs` for the rules these types keep.
#[derive(Serialize)]
struct DescriptionShape {
    id: String,
    name: String,
}

const DUPLICATE: &str = "A description with this name already exists";

/// Every description, as `(id, name)`.
///
/// Case folding happens in Rust, not SQL, and that is the whole reason this
/// function exists. **SQLite folds ASCII only** — `lower('CAFÉ')` is `'cafÉ'`,
/// and `COLLATE NOCASE` is no better; measured, not assumed. Postgres's
/// `mode: 'insensitive'` folded full Unicode, so leaving the comparison in SQL
/// would have quietly made "CAFÉ" and "café" two merchants on SQLite where they
/// were one before. `str::to_lowercase` folds the whole of Unicode.
///
/// Reading the table to compare it is affordable because it is a list of
/// merchant names — tens to hundreds of rows, and the alternative is a stored
/// fold column that has to be kept in step with the name forever.
async fn all_names(pool: &SqlitePool) -> Result<Vec<(String, String)>, ApiError> {
    Ok(sqlx::query!(
        r#"SELECT "id" AS "id!", "name" AS "name!"
             FROM "TransactionDescription" ORDER BY "name" ASC"#
    )
    .fetch_all(pool)
    .await?
    .into_iter()
    .map(|r| (r.id, r.name))
    .collect())
}

/// Is this name already taken, ignoring case and one id?
async fn name_taken(pool: &SqlitePool, name: &str, except: &str) -> Result<bool, ApiError> {
    let wanted = name.to_lowercase();
    Ok(all_names(pool)
        .await?
        .iter()
        .any(|(id, n)| id != except && n.to_lowercase() == wanted))
}

async fn read(pool: &SqlitePool, id: &str) -> Result<DescriptionShape, ApiError> {
    let r = sqlx::query!(
        r#"SELECT "id" AS "id!", "name" AS "name!"
             FROM "TransactionDescription" WHERE "id" = ?"#,
        id
    )
    .fetch_optional(pool)
    .await?
    .ok_or_else(|| ApiError::not_found("Description"))?;
    Ok(DescriptionShape {
        id: r.id,
        name: r.name,
    })
}

pub async fn list(pool: &SqlitePool, p: &Path<'_>) -> Result<Response, ApiError> {
    // Filtered in Rust for the same reason as `name_taken`, plus one more: a
    // substring match in SQL means LIKE, and LIKE treats `%` and `_` as
    // wildcards — so a search for "50%" would match every description unless
    // the input were escaped. `str::contains` has no metacharacters.
    let search = p
        .query("search")
        .filter(|s| !s.is_empty())
        .map(str::to_lowercase);

    let rows = all_names(pool).await?;
    Ok(Response::ok(
        rows.into_iter()
            .filter(|(_, name)| match &search {
                Some(q) => name.to_lowercase().contains(q),
                None => true,
            })
            .map(|(id, name)| DescriptionShape { id, name })
            .collect::<Vec<_>>(),
    ))
}

#[derive(Deserialize, Default)]
#[serde(default)]
struct NameBody {
    name: String,
}

pub async fn create(pool: &SqlitePool, body: Option<Value>) -> Result<Response, ApiError> {
    let b: NameBody = crate::body_of(body)?;
    let name = b.name.trim().to_string();
    if name.is_empty() {
        return Err(crate::recurring::required("name"));
    }
    if name_taken(pool, &name, "").await? {
        return Err(ApiError::conflict(DUPLICATE));
    }

    let id = cuid();
    let now = now_iso();
    sqlx::query!(
        r#"INSERT INTO "TransactionDescription" ("id","name","createdAt")
           VALUES (?,?,?)"#,
        id,
        name,
        now
    )
    .execute(pool)
    .await?;
    Ok(Response::created(DescriptionShape { id, name }))
}

/// Rename a description, and every transaction filed under it.
pub async fn rename(
    pool: &SqlitePool,
    id: &str,
    body: Option<Value>,
) -> Result<Response, ApiError> {
    let b: NameBody = crate::body_of(body)?;
    let name = b.name.trim().to_string();
    if name.is_empty() {
        return Err(crate::recurring::required("name"));
    }
    read(pool, id).await?;
    // Excluding self, so re-saving a name unchanged — or changing only its
    // capitalisation — is allowed rather than a conflict with itself.
    if name_taken(pool, &name, id).await? {
        return Err(ApiError::conflict(DUPLICATE));
    }

    sqlx::query!(
        r#"UPDATE "TransactionDescription" SET "name" = ? WHERE "id" = ?"#,
        name,
        id
    )
    .execute(pool)
    .await?;

    retitle(pool, id, &name, None).await?;
    Ok(Response::ok(DescriptionShape {
        id: id.to_string(),
        name,
    }))
}

/// Point every transaction on `from_description` at `name`, and optionally move
/// it to another description.
///
/// One `ledger_update` per row rather than a bulk UPDATE — see the module note.
/// The connection is scoped so it is released before anything else asks the
/// pool for one; the pool holds exactly one.
async fn retitle(
    pool: &SqlitePool,
    from_description: &str,
    name: &str,
    move_to: Option<&str>,
) -> Result<usize, ApiError> {
    let ids = sqlx::query_scalar!(
        r#"SELECT "id" AS "id!" FROM "Transaction" WHERE "descriptionId" = ?"#,
        from_description
    )
    .fetch_all(pool)
    .await?;

    let mut conn = pool.acquire().await?;
    for tx in &ids {
        ledger_update(
            &mut conn,
            tx,
            &LedgerUpdate {
                name: Some(name.to_string()),
                description_id: move_to.map(|d| Some(d.to_string())),
                ..Default::default()
            },
        )
        .await
        .map_err(ApiError::from)?;
    }
    Ok(ids.len())
}

#[derive(Deserialize, Default)]
#[serde(default)]
struct MergeBody {
    #[serde(rename = "sourceIds")]
    source_ids: Vec<String>,
    #[serde(rename = "targetId")]
    target_id: String,
}

/// Fold several descriptions into one.
pub async fn merge(pool: &SqlitePool, body: Option<Value>) -> Result<Response, ApiError> {
    let b: MergeBody = crate::body_of(body)?;
    let target = read(pool, &b.target_id)
        .await
        .map_err(|_| ApiError::not_found("Target description"))?;
    let name = target.name.clone();

    // Merging a description into itself is a no-op, not an error — a UI that
    // selects a range including the target should not have to filter it out.
    let sources: Vec<String> = b
        .source_ids
        .iter()
        .filter(|id| **id != b.target_id)
        .cloned()
        .collect();
    if sources.is_empty() {
        return Ok(Response::ok(target));
    }

    // All-or-nothing on existence: a partial merge would move some rows and
    // leave the caller unable to tell which.
    for id in &sources {
        let n = sqlx::query_scalar!(
            r#"SELECT count(*) FROM "TransactionDescription" WHERE "id" = ?"#,
            id
        )
        .fetch_one(pool)
        .await?;
        if n == 0 {
            return Err(ApiError::new(
                404,
                "One or more source descriptions not found",
            ));
        }
    }

    for id in &sources {
        retitle(pool, id, &name, Some(&b.target_id)).await?;
        // Safe now: nothing references it.
        sqlx::query!(r#"DELETE FROM "TransactionDescription" WHERE "id" = ?"#, id)
            .execute(pool)
            .await?;
    }

    Ok(Response::ok(target))
}

#[derive(Deserialize, Default)]
#[serde(default)]
struct MergeIntoBody {
    #[serde(rename = "mergeId")]
    merge_id: String,
}

/// The same fold, addressed from the surviving description.
pub async fn merge_into(
    pool: &SqlitePool,
    id: &str,
    body: Option<Value>,
) -> Result<Response, ApiError> {
    let b: MergeIntoBody = crate::body_of(body)?;
    if id == b.merge_id {
        return Err(ApiError::bad_request(
            "Cannot merge a description into itself",
        ));
    }
    let target = read(pool, id)
        .await
        .map_err(|_| ApiError::not_found("Target description"))?;
    let name = target.name.clone();
    read(pool, &b.merge_id)
        .await
        .map_err(|_| ApiError::not_found("Source description"))?;

    retitle(pool, &b.merge_id, &name, Some(id)).await?;
    sqlx::query!(
        r#"DELETE FROM "TransactionDescription" WHERE "id" = ?"#,
        b.merge_id
    )
    .execute(pool)
    .await?;

    Ok(Response::ok(target))
}

/// Remove a description nothing is filed under.
///
/// Refused while transactions reference it, with the count, because the answer
/// to "why can't I delete this" is the number — and merging is the operation
/// they actually want.
pub async fn delete(pool: &SqlitePool, id: &str) -> Result<Response, ApiError> {
    read(pool, id).await?;
    let n = sqlx::query_scalar!(
        r#"SELECT count(*) FROM "Transaction" WHERE "descriptionId" = ?"#,
        id
    )
    .fetch_one(pool)
    .await?;
    if n > 0 {
        return Err(ApiError::conflict(format!(
            "Cannot delete: {n} transaction(s) still reference this description"
        )));
    }
    sqlx::query!(r#"DELETE FROM "TransactionDescription" WHERE "id" = ?"#, id)
        .execute(pool)
        .await?;
    Ok(Response::no_content())
}

File: apps/api/src/routes/descriptions.ts
import { createRoute } from '@hono/zod-openapi';
import { z } from 'zod';
import { prisma, Prisma } from '@budget-tracker/db';
import {
  DescriptionSchema,
  CreateDescriptionSchema,
  RenameDescriptionSchema,
  MergeDescriptionsSchema,
} from '@budget-tracker/core';
import { ErrorSchema, createRouter } from '../lib/errors.js';
import { ledgerUpdate } from '../lib/lifecycle/index.js';

const app = createRouter();

// ─── GET / ───

const listDescriptionsRoute = createRoute({
  method: 'get',
  path: '/',
  tags: ['Descriptions'],
  summary: 'List all transaction descriptions',
  request: {
    query: z.object({
      search: z.string().optional(),
    }),
  },
  responses: {
    200: {
      content: { 'application/json': { schema: z.array(DescriptionSchema) } },
      description: 'List of descriptions',
    },
  },
});

app.openapi(listDescriptionsRoute, async (c) => {
  const { search } = c.req.valid('query');

  const where: Prisma.TransactionDescriptionWhereInput = search
    ? { name: { contains: search, mode: 'insensitive' } }
    : {};

  const descriptions = await prisma.transactionDescription.findMany({
    where,
    select: { id: true, name: true },
    orderBy: { name: 'asc' },
  });

  return c.json(descriptions, 200);
});

// ─── POST / ───

const createDescriptionRoute = createRoute({
  method: 'post',
  path: '/',
  tags: ['Descriptions'],
  summary: 'Create a transaction description',
  request: {
    body: { content: { 'application/json': { schema: CreateDescriptionSchema } } },
  },
  responses: {
    201: {
      content: { 'application/json': { schema: DescriptionSchema } },
      description: 'Description created',
    },
    400: {
      content: { 'application/json': { schema: ErrorSchema } },
      description: 'Bad Request',
    },
    409: {
      content: { 'application/json': { schema: ErrorSchema } },
      description: 'Conflict',
    },
  },
});

app.openapi(createDescriptionRoute, async (c) => {
  const { name } = c.req.valid('json');

  // Case-insensitive uniqueness check
  const existing = await prisma.transactionDescription.findFirst({
    where: { name: { equals: name, mode: 'insensitive' } },
  });

  if (existing) {
    return c.json({ error: 'A description with this name already exists' }, 409);
  }

  const description = await prisma.transactionDescription.create({
    data: { name },
    select: { id: true, name: true },
  });

  return c.json(description, 201);
});

// ─── PUT /:id ───

const renameDescriptionRoute = createRoute({
  method: 'put',
  path: '/{id}',
  tags: ['Descriptions'],
  summary: 'Rename a transaction description',
  request: {
    params: z.object({ id: z.string() }),
    body: { content: { 'application/json': { schema: RenameDescriptionSchema } } },
  },
  responses: {
    200: {
      content: { 'application/json': { schema: DescriptionSchema } },
      description: 'Description renamed',
    },
    400: {
      content: { 'application/json': { schema: ErrorSchema } },
      description: 'Bad Request',
    },
    404: {
      content: { 'application/json': { schema: ErrorSchema } },
      description: 'Not Found',
    },
    409: {
      content: { 'application/json': { schema: ErrorSchema } },
      description: 'Conflict',
    },
  },
});

app.openapi(renameDescriptionRoute, async (c) => {
  const { id } = c.req.valid('param');
  const { name } = c.req.valid('json');

  // Case-insensitive uniqueness check (exclude self)
  const existing = await prisma.transactionDescription.findFirst({
    where: {
      name: { equals: name, mode: 'insensitive' },
      id: { not: id },
    },
  });

  if (existing) {
    return c.json({ error: 'A description with this name already exists' }, 409);
  }

  try {
    const description = await prisma.transactionDescription.update({
      where: { id },
      data: { name },
      select: { id: true, name: true },
    });

    // Update all linked transactions through the ledger gate
    const txs = await prisma.transaction.findMany({
      where: { descriptionId: id },
      select: { id: true },
    });
    for (const tx of txs) {
      await ledgerUpdate(tx.id, { name });
    }

    return c.json(description, 200);
  } catch (err) {
    if (err instanceof Prisma.PrismaClientKnownRequestError) {
      if (err.code === 'P2025') return c.json({ error: 'Description not found' }, 404);
      if (err.code === 'P2002')
        return c.json({ error: 'A description with this name already exists' }, 409);
    }
    throw err;
  }
});

// ─── POST /merge ───

const mergeDescriptionsRoute = createRoute({
  method: 'post',
  path: '/merge',
  tags: ['Descriptions'],
  summary: 'Merge multiple descriptions into one',
  request: {
    body: { content: { 'application/json': { schema: MergeDescriptionsSchema } } },
  },
  responses: {
    200: {
      content: { 'application/json': { schema: DescriptionSchema } },
      description: 'Merge completed, returns the target description',
    },
    400: {
      content: { 'application/json': { schema: ErrorSchema } },
      description: 'Bad Request',
    },
    404: {
      content: { 'application/json': { schema: ErrorSchema } },
      description: 'Not Found',
    },
  },
});

app.openapi(mergeDescriptionsRoute, async (c) => {
  const { sourceIds, targetId } = c.req.valid('json');

  // Validate target exists
  const target = await prisma.transactionDescription.findUnique({
    where: { id: targetId },
    select: { id: true, name: true },
  });

  if (!target) {
    return c.json({ error: 'Target description not found' }, 404);
  }

  // Filter out targetId from sourceIds to avoid self-merge
  const filteredSourceIds = sourceIds.filter((id) => id !== targetId);

  if (filteredSourceIds.length === 0) {
    return c.json(target, 200);
  }

  // Verify all source descriptions exist
  const sources = await prisma.transactionDescription.findMany({
    where: { id: { in: filteredSourceIds } },
    select: { id: true },
  });

  if (sources.length !== filteredSourceIds.length) {
    return c.json({ error: 'One or more source descriptions not found' }, 404);
  }

  // Move all transactions from sources to target through the ledger gate
  const txs = await prisma.transaction.findMany({
    where: { descriptionId: { in: filteredSourceIds } },
    select: { id: true },
  });
  for (const t of txs) {
    await ledgerUpdate(t.id, { descriptionId: targetId, name: target.name });
  }

  // Delete source descriptions (safe now that no transactions reference them)
  await prisma.transactionDescription.deleteMany({
    where: { id: { in: filteredSourceIds } },
  });

  return c.json(target, 200);
});

// ─── POST /:id/merge ───

const mergeIntoDescriptionRoute = createRoute({
  method: 'post',
  path: '/{id}/merge',
  tags: ['Descriptions'],
  summary: 'Merge another description into this one',
  request: {
    params: z.object({ id: z.string() }),
    body: { content: { 'application/json': { schema: z.object({ mergeId: z.string() }) } } },
  },
  responses: {
    200: {
      content: { 'application/json': { schema: DescriptionSchema } },
      description: 'Merge completed, returns the target description',
    },
    400: {
      content: { 'application/json': { schema: ErrorSchema } },
      description: 'Bad Request',
    },
    404: {
      content: { 'application/json': { schema: ErrorSchema } },
      description: 'Not Found',
    },
  },
});

app.openapi(mergeIntoDescriptionRoute, async (c) => {
  const { id } = c.req.valid('param');
  const { mergeId } = c.req.valid('json');

  if (id === mergeId) {
    return c.json({ error: 'Cannot merge a description into itself' }, 400);
  }

  // Validate target exists
  const target = await prisma.transactionDescription.findUnique({
    where: { id },
    select: { id: true, name: true },
  });

  if (!target) {
    return c.json({ error: 'Target description not found' }, 404);
  }

  // Validate source exists
  const source = await prisma.transactionDescription.findUnique({
    where: { id: mergeId },
    select: { id: true },
  });

  if (!source) {
    return c.json({ error: 'Source description not found' }, 404);
  }

  // Move all transactions from source to target through the ledger gate
  const txs = await prisma.transaction.findMany({
    where: { descriptionId: mergeId },
    select: { id: true },
  });
  for (const t of txs) {
    await ledgerUpdate(t.id, { descriptionId: id, name: target.name });
  }

  // Delete the source description (safe now that no transactions reference it)
  await prisma.transactionDescription.delete({
    where: { id: mergeId },
  });

  return c.json(target, 200);
});

// ─── DELETE /:id ───

const deleteDescriptionRoute = createRoute({
  method: 'delete',
  path: '/{id}',
  tags: ['Descriptions'],
  summary: 'Delete a description (only if no transactions reference it)',
  request: {
    params: z.object({ id: z.string() }),
  },
  responses: {
    204: { description: 'Deleted' },
    404: {
      content: { 'application/json': { schema: ErrorSchema } },
      description: 'Not Found',
    },
    409: {
      content: { 'application/json': { schema: ErrorSchema } },
      description: 'Conflict — transactions still reference this description',
    },
  },
});

app.openapi(deleteDescriptionRoute, async (c) => {
  const { id } = c.req.valid('param');

  // Check if description exists
  const description = await prisma.transactionDescription.findUnique({
    where: { id },
    select: { id: true },
  });

  if (!description) {
    return c.json({ error: 'Description not found' }, 404);
  }

  // Check if any transactions reference it
  const txCount = await prisma.transaction.count({
    where: { descriptionId: id },
  });

  if (txCount > 0) {
    return c.json(
      { error: `Cannot delete: ${txCount} transaction(s) still reference this description` },
      409,
    );
  }

  await prisma.transactionDescription.delete({ where: { id } });
  return c.body(null, 204);
});

export default app;
